src/counts.jl

```diff
@@ -19,21 +19,21 @@ end
     addcounts!(r, x, levels::UnitRange{<:Int}, [wv::AbstractWeights])
 
 Add the number of occurrences in `x` of each value in `levels` to an existing
-array `r`. If a weighting vector `wv` is specified, the sum of weights is used
-rather than the raw counts.
+array `r`.
+
+If a weighting vector `wv` is specified, the sum of weights is used rather than the
+raw counts.
 """
 function addcounts!(r::AbstractArray, x::IntegerArray, levels::IntUnitRange)
-    # add counts of integers from x to r
+    # add counts of integers from x that fall within levels to r
 
-    k = length(levels)
-    length(r) == k || throw(DimensionMismatch())
+    checkbounds(r, axes(levels)...)
```

Comment:
I think
```suggestion
    (firstindex(r) == 1) && length(r) == k) || throw(DimensionMismatch())
```
better expresses the intention in the original code. Also maybe this condition should be noted in the documentation.

Reply:
Yes, the previous version errored when `r` was larger than `levels` and if we keep that restriction, it should be noted in the documentation.

However, it feels more natural to me _not_ to keep that restriction, especially since it is not documented. So long as what we are adding to `r` fits into `r`, this method will work and other entries of `r` will not be affected. This looser restriction and guarantee not to mess with other entries of `r` is implicit and need not be mentioned in the docstring.

Reply:
The issue is broader. With your code if `r` is `OffsetArray` then indices from `1` to length(levels)` must be valid in it. But then you are adding values to `r` in indices starting from `1`, which in general does not have to be first element of `r`. The docstring does not specify where the elements are put into `r`. There can be two options:
* starting from index `1` or
* starting from beginning of `r`

That is why I propose to add `firstindex(r)==1` condition which resolves this issue as then both options are the same.
I agree that we could allow `r` to be longer than `length(levels)` (this is a change, but I think it is acceptable and makes sense).

Reply:
Option 1 sounds fine, the way `r` is used should have been explained better anyway in the current version (one could have imagined that the value of each level was used as to index into `r` rather than starting at 1). Allowing `r` to be longer than `length(levels)` sounds OK too as long as it's documented.

Reply:
> then indices from `1` to `length(levels)` must be valid in it

@bkamins this is actually already checked by `checkbounds(r, axes(levels)...)` above, since the latter is `OneTo(...)`. But given that I raised similar issue two days ago (in comments just below) there might be some merit in documenting it ;)

Reply:
Yes - it is chcecked. My point is that it should be documented.

Reply:
@kalmarek 
> But given that I raised similar issue two days ago (in comments just below) there might be some merit in documenting it ;)

Could you link that issue? I'm not seeing it and want to make sure the documentation I added addresses it.

Reply:
For completeness, here is the link: https://github.com/JuliaStats/StatsBase.jl/pull/722#discussion_r874519428